Reopen user case when mobile worker is reactivated

NoahCarnahan opened the pull request:
http://manage.dimagi.com/default.asp?187418
Best reviewed commit-by-commit.
@kaapstorm cc @nickpell

NoahCarnahan committed:
[refactor] Break sync_user_case into several functions

NoahCarnahan committed:
Reopen closed user cases if user is unarchived

dimagimon commented on corehq/apps/callcenter/utils.py:
```diff
@@ -116,6 +92,51 @@ def valid_element_name(name):
+            return False
+
+
+    # remove any keys that aren't valid XML element names
```
E303 too many blank lines (2)

NoahCarnahan committed:
Update utils.py

nickpell commented:
looks good to merge pending tests

NoahCarnahan commented:
There's an unhanded edge case that I realized needs addressing. Hold off on merge until I get that straightened out.

kaapstorm commented:
:+1:

NoahCarnahan committed:
Fix sync_user_case logic

- If a closed case is updated, don't submit a new close block
- If a case is update and closed in one save. Submit update and close
  blocks separately.
- When reopening a case, don't assume last case action is a close.
Also breaks sync_user_case into smaller methods.

NoahCarnahan committed:
[refactor] Group methods in _UserCaseHelper class

NoahCarnahan committed:
Merge branch 'reopen-usercase' of github.com:dimagi/commcare-hq into reopen-usercase

NoahCarnahan commented:
@kaapstorm I added a commit (15793f9) that changes the logic a bit more. It would be great if you could confirm that this approach seems sound.
- No longer submit close actions when updating already closed cases. Previously, an update to a closed case (which could happen if an archived user is modified) would also submit another close action. This made it challenging to archive the form that closes a case, which is the technique I'm using to re-open closed user cases.
- Submit two separate forms with update and close blocks if a case is updated and closed at the same time. This is also necessary for the reopening workflow.

kaapstorm commented on corehq/apps/callcenter/utils.py:
```diff
@@ -52,68 +106,64 @@ def sync_user_case(commcare_user, case_type, owner_id, case=None):
         domain = commcare_user.project
+        fields = _get_user_case_fields(commcare_user)
+        case = case or get_case_by_domain_hq_user_id(domain.name, commcare_user._id, case_type)
+        close = commcare_user.to_be_deleted() or not commcare_user.is_active
```
Does this mean that when a user is deleted, their user case is closed, not marked as deleted? Is this correct? (It is different from the behaviour of normal cases.)

NoahCarnahan replied:
I believe that's right. When a user is deactivated, the case is indeed closed, but not deleted. It's hard to tell in the diff, but that's not actually new behavior in this PR. This particular line existed beforehand.

kaapstorm replied:
:+1:

kaapstorm commented:
This all looks great (assuming previous question is correct behaviour). 

Could you add a couple of tests showing what 060a87c and 15793f9 do? [CallCenterTestOpenedClosed][1] might be a useful place to add them. 

  [1]: https://github.com/dimagi/commcare-hq/blob/c2ccefafdd1861c0571dd2d505cbc731a3d5dd8f/corehq/apps/callcenter/tests/test_indicators.py#L491-L491

NoahCarnahan committed:
Allow usercase to be updated and re-opened in same user model save

NoahCarnahan committed:
Add tests for user case sync issues

kaapstorm committed:
Merge branch 'master' into reopen-usercase

kaapstorm commented:
Thanks for the tests, @NoahCarnahan. I took the liberty to merge master to see whether Travis comes back green this time.

NoahCarnahan commented:
Awesome, thanks Norman!
